### DISCUSSION TITLE
Speed up "reading" times with icon overrides present?

### DISCUSSION THREAD

**Discussion by J-animA** [squidfunk/mkdocs-material#2975] [Q&A]
Hello Everyone,

I recently updated one of my docs sites to mkdocs-material 7.2.5 (issue was also seen in 7.2.4) and MkDocs 1.2.2 and followed the instructions for extending the theme and adding custom icons (for use with the theme toggles) https://squidfunk.github.io/mkdocs-material/setup/changing-the-logo-and-icons/#additional-icons

Relevant lines in `mkdocs.yml`:

```yaml
markdown_extensions:
  - pymdownx.emoji:
      emoji_index: !!python/name:materialx.emoji.twemoji
      emoji_generator: !!python/name:materialx.emoji.to_svg
      options:
        custom_icons:
          - overrides/.icons

theme:
  name: material
  custom_dir: overrides
  palette:
    - scheme: slate
      primary: deep purple
      accent: purple
      toggle:
        icon: custom/MOON
        name: Switch to dark mode
    - scheme: default
      primary: deep purple
      accent: purple
      toggle:
        icon: custom/SUN
        name: Switch to light mode
 ```

When these overrides are in place, it takes a very long time for the site to build (specifically with `mkdocs serve` but the same issue occurs with `mkdocs build` as well):

```bash
$ mkdocs serve
INFO     -  Building documentation...
INFO     -  Cleaning site directory
...
...
INFO     -  Documentation built in 121.92 seconds
```

With the overrides removed it takes a "normal" amount of time to build and serve:
```bash
$ mkdocs serve
INFO     -  Building documentation...
INFO     -  Cleaning site directory
...
...
INFO     -  Documentation built in 10.86 seconds
INFO     -  [12:53:10] Serving on http://127.0.0.1:8000/
```

considering there is a checklist item to disable overrides when reporting a bug/issue, I'm assuming this is expected in some fashion, but my question is: is there a way to determine _why_ there is such a slowdown? For reference, there are a whole 2 `.svg` files in the custom directory. It's structure is as follows:

```yaml
.
├─ overrides/
│  └─ .icons/
│     └─ custom/
│        └─ SUN.svg
|        └─ MOON.svg
└─ mkdocs.yml
```

I'd appreciate any info.

**Comment by squidfunk** [comment]
> ... considering there is a checklist item to disable overrides when reporting a bug/issue, I'm assuming this is expected in some fashion, but my question is: ...

Correct, we cannot control what users do with this project, and plugins or extension may cause significant slowdown, which is why, in general, we only offer support for issues that are not related to customization.

> ... is there a way to determine why there is such a slowdown?

Nope, you'd have to narrow it down by commenting out your customizations one-by-one. If you've isolated the cause, we can look into it, but it looks like it's more likely related to an upstream dependency of Material for MkDocs.

**Reply by J-animA** [reply]
Yes it is indeed the icon override causing the issue as the site builds/serves normally when these are commented out of the `mkdocs.yml`

I'll do some further testing with another icon pack like `Bootstrap` as mentioned in the docs and see if it occurs with that as well.

Thank you!

**Reply by J-animA** [reply]
Yep, looks like it occurs with bootstrap as well, so I think the theory of this being related to an upstream dependency is plausible, or maybe it's a quirk of MkDocs itself.

If it helps, it looks like the actual bottleneck is when mkdocs is "reading" the documentation. The actual build portion is pretty fast. It's difficult to get across in a code block but trust me when I say the "reading" entries here are what's taking time. The "copying" and "building" log outputs are speedy as usual:

Truncated output of `mkdocs serve -v`:
```bash
DEBUG    -  Reading markdown pages.
DEBUG    -  Reading: index.md
DEBUG    -  Reading: about.md
DEBUG    -  Reading: contact/cpeteam.md
DEBUG    -  Reading: contact/geo-pres.md
...
...
DEBUG    -  Reading: rdkb_docs/client-connectivity.md
DEBUG    -  Reading: rdkb_docs/consoles.md
DEBUG    -  Reading: rdkb_docs/firmware_request.md
DEBUG    -  Reading: rdkb_docs/landing.md
DEBUG    -  Reading: rdkb_docs/mesh_cpe_triage.md
DEBUG    -  Copying static assets.
DEBUG    -  Copying media file: 'files/HsdReLogos.zip'
DEBUG    -  Copying media file: 'files/HsdReLogos_nosubtitle.zip'
DEBUG    -  Copying media file: 'files/bookmarks.html'
DEBUG    -  Copying media file: 'files/cbr_tcpdump.tar.gz'
DEBUG    -  Copying media file: 'files/logo.JPG'
DEBUG    -  Copying media file: 'files/tcpdump-cxb3'
...
...
DEBUG    -  Building theme template: sitemap.xml
DEBUG    -  Gzipping template: sitemap.xml
DEBUG    -  Building theme template: 404.html
DEBUG    -  Building markdown pages.
DEBUG    -  Building page index.md
DEBUG    -  Building page about.md
DEBUG    -  Building page contact/cpeteam.md
DEBUG    -  Building page contact/geo-pres.md
DEBUG    -  Building page contact/team.md
...
...
INFO     -  Documentation built in 220.72 seconds
```

With the overrides commented out in `mkdocs.yml`:

```
INFO     -  Documentation built in 9.10 seconds
```

**Reply by J-animA** [reply]
Updated the discussion title to better reflect the issue

**Reply by squidfunk** [reply]
Thanks for the info. @facelessuser might this be related to the [extensions package](https://github.com/facelessuser/mkdocs-material-extensions)? Any idea?

**Reply by squidfunk** [reply]
@jaredlmiller what happens if you specify the overrides directory but leave it empty?

**Reply by J-animA** [reply]
@squidfunk leaving the following under the `markdown_extensions` block with no subdirectory within results in the same issue:

```yaml
      options:
        custom_icons:
          - overrides/.icons
```

With Override (only 10-12seconds for file size sake, but this gets slower and slower as it moves on):
![with](https://github.com/jaredlmiller/images/blob/main/with_override.gif?raw=true)

Override Commented:
![commented](https://github.com/jaredlmiller/images/blob/main/override_commented.gif?raw=true)

**Reply by squidfunk** [reply]
Okay, then I'm almost sure that it is either related to [Python Markdown Extensions](https://github.com/facelessuser/pymdown-extensions) or [MkDocs](https://github.com/mkdocs/mkdocs) itself. I'd recommend posting on the [discussion board](https://github.com/facelessuser/pymdown-extensions/discussions) of the former first, before taking it to MkDocs.

**Reply by facelessuser** [reply]
I'm not sure what could be done if the Material emoji extension is to blame. When the plugin is loaded for the first time, it should index all the icons. It should only do this when the plugin is loaded for the first time. I'm not sure if MkDocs is re-instantiating the plugins each time or what. If it is re-instantiating it multiple times, then that could be the problem. If it was shown that we were indexing more than once when the plugin was only instantiated once, then we could probably fix that, but outside of that, I'm not sure what the problem would be.

Also, I'm not sure what the "Reading" step is doing in MkDocs. Is it running the Markdown processor in this step?

**Reply by facelessuser** [reply]
I'll run some basic tests with the latest `pymdown-extensions` and the latest `mkdocs-material-extensions`. If I don't see the icon list compounding or indexing more than expected, I'm not sure the problem is on my side.

**Reply by facelessuser** [reply]
From what I can see, icons are indexed on every page, but only once on each page. The number of paths doesn't inflate on accident, and the number of icons and aliases aren't inflating on accident either. The question is whether indexing is getting run only on new instantiations or whether they are accidentally re-running on new calls to old instantiated instances.

```
DEBUG    -  Reading: about/changelog.md
---Paths to search---
['docs/theme/my_icons', '/usr/local/lib/python3.9/site-packages/material/.icons']
---indexed---
11252
1762
DEBUG    -  Reading: about/contributing.md
---Paths to search---
['docs/theme/my_icons', '/usr/local/lib/python3.9/site-packages/material/.icons']
---indexed---
11252
1762
DEBUG    -  Reading: about/license.md
---Paths to search---
['docs/theme/my_icons', '/usr/local/lib/python3.9/site-packages/material/.icons']
---indexed---
11252
1762
```

**Reply by facelessuser** [reply]
From what I can tell, emoji will only run the custom icon indexing when the new class is instantiated (see `_set_index` which is only run under `__init__`):

```py
class EmojiPattern(InlineProcessor):
    """Return element of type `tag` with a text attribute of group(2) of an `InlineProcessor`."""

    def __init__(self, pattern, config, md):
        """Initialize."""

        InlineProcessor.__init__(self, pattern, md)

        title = config['title']
        alt = config['alt']
        self.options = config['options']
        self._set_index(config["emoji_index"])
        self.unicode_alt = alt in UNICODE_ALT
        self.encoded_alt = alt == UNICODE_ENTITY
        self.remove_var_sel = config['remove_variation_selector']
        self.title = title if title in VALID_TITLE else NO_TITLE
        self.generator = config['emoji_generator']

    def _set_index(self, index):
        """Set the index."""

        if len(inspect.getfullargspec(index).args):
            self.emoji_index = index(self.options, self.md)
        else:
            warnings.warn(MSG_INDEX_WARN, util.PymdownxDeprecationWarning)
            self.emoji_index = index()
```

So it seems like a new instance of Python Markdown is created for each file. If you are scanning a large custom folder of icons, I imagine you will increase the time even more.

Now, this is all based on the latest versions. If you are using an old version of `mkdocs-material-extensions` there was a bug where the index would inflate bigger and bigger, but this is fixed in the latest iteration.
